File: grind75/week1/054-Min-Stack-Medium/054-Min-Stack-Medium.cpp
// Problem link: https://leetcode.com/problems/min-stack/
class MinStack {
public:
    std::stack<int> stack;
    std::stack<int> minStack;

    MinStack() {
        
    }
    
    void push(int val) {
        stack.push(val);
        if(minStack.empty())
        {
            minStack.push(val);
        }
        else
        {
            minStack.push(std::min(val, minStack.top()));
        }
    }
    
    void pop() {
        stack.pop();
        minStack.pop();
    }
    
    int top() {
        return stack.top();
    }
    
    int getMin() {
        return minStack.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: grind75/week1/054-Min-Stack-Medium/054-Min-Stack-Medium.py
# Question link: https://leetcode.com/problems/min-stack/
# This one can be a little tricky, I'm not gonna explain it through comments, just watch this video, it's more concise
# yt: https://www.youtube.com/watch?v=qkLl7nAwDPo

class MinStack:

    def __init__(self):
        self.stack = []
        self.minStack = []
        

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.minStack:
            self.minStack.append(val)
        else:
            val = min(val, self.minStack[-1])
            self.minStack.append(val)
        

    def pop(self) -> None:
        self.stack.pop()
        self.minStack.pop()
        

    def top(self) -> int:
        return self.stack[-1]
        

    def getMin(self) -> int:
       return self.minStack[-1] 



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
